core/src/main/java/hudson/util/RobustMapConverter.java

```diff
@@ -73,7 +103,19 @@ private Object read(HierarchicalStreamReader reader, UnmarshallingContext contex
         }
         reader.moveDown();
         try {
-            return readBareItem(reader, context, map);
+            var object = readBareItem(reader, context, map);
+            try {
+                if (expectedType != null) {
+                    // When possible, disallow invalid keys and values.
+                    expectedType.cast(object);
+                }
+                return object;
+            } catch (ClassCastException e) {
```

Comment:
This definitely looks like it should be using https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Class.html#isInstance(java.lang.Object)

Reply:
Originally this was to be able to throw the `ClassCastException` directly, but yes, now that an explicit `ConversionException` is needed to record contextual information anyway, this can probably be simplified.

Reply:
[df2ac81](https://github.com/jenkinsci/jenkins/pull/9727/commits/df2ac81257510f5f0ac715d2bf5036b20b2ffaf7)